### PR TITLE
DAOS-7487 test: Extend autotest with more smoke capabilities

### PR DESCRIPTION
**Pull request by chrisphoffman**
Extended autotest to include additional functionality:

-better diagnoses, more particularly when the agent isn't running or one
cannot connect to the server (e.g. network issue)
-use the new API to auto-select the object class
-create extra containers with RF=1 and RF=2

Signed-off-by: Christopher Hoffman <christopherx.hoffman@intel.com>

### CODE REVIEW

**Comment by kprantis** [comment]
I think this needs rebased as there was a regression on master this morning.

File: src/utils/daos_autotest.c [hunk 1]
```diff
@@ -32,6 +33,11 @@ clock_t	end;
 
 /** generated container UUID */
 uuid_t		cuuid;
+
+/** generate container UUID for aux cont */
+uuid_t		cuuid2;
+uuid_t		cuuid3;
```
Comment (johannlombardi) [inline src/utils/daos_autotest.c:39]: We could just use labels now for all the containers.

File: src/utils/daos_autotest.c [hunk 2]
```diff
@@ -112,6 +118,32 @@ step_init(void)
 	fprintf(autotest_ap->outstream, "Status Time(sec) Comment\033[0m\n");
 }
 
+static int
+attachinfo(void)
+{
+	char	*command = "sudo daos_agent dump-attachinfo";
```
Comment (johannlombardi) [inline src/utils/daos_autotest.c:124]: sudo? Why is that needed?

File: src/utils/daos_autotest.c [hunk 3]
```diff
@@ -156,6 +188,36 @@ ccreate(void)
...
 	}
 
+	/** Create container with RF=1 */
+	daos_prop_t	*prop;
+
+	prop = daos_prop_alloc(1);
+	prop->dpp_entries[0].dpe_type = DAOS_PROP_CO_REDUN_FAC;
+	prop->dpp_entries[0].dpe_val = DAOS_PROP_CO_REDUN_RF1;
+
+	uuid_generate(cuuid2);
```
Comment (johannlombardi) [inline src/utils/daos_autotest.c:198]: New API does not need uuid to be provided.

File: src/utils/daos_autotest.c [hunk 4]
```diff
@@ -156,6 +188,36 @@ ccreate(void)
...
+
+	rc = daos_cont_create(poh, cuuid2, prop, NULL);
```
Comment (johannlombardi) [inline src/utils/daos_autotest.c:200]: cuuid2 should just be a string here. e.g. "autotest-$hostname-$pid"

File: src/utils/daos_autotest.c [hunk 5]
```diff
@@ -156,6 +188,36 @@ ccreate(void)
...
+	}
+
+	/** Create container with RF=2 */
+	daos_prop_t	*prop2;
+
+	prop2 = daos_prop_alloc(1);
+	prop2->dpp_entries[0].dpe_type = DAOS_PROP_CO_REDUN_FAC;
+	prop2->dpp_entries[0].dpe_val = DAOS_PROP_CO_REDUN_RF2;
+
+	uuid_generate(cuuid3);
```
Comment (johannlombardi) [inline src/utils/daos_autotest.c:213]: same here.

File: src/utils/daos_autotest.c [hunk 6]
```diff
@@ -479,7 +541,7 @@ kv_insert128(void)
 	int		rc;
 
 	new_oid();
-	daos_obj_generate_oid(coh, &oid, DAOS_OF_KV_FLAT, OC_SX, 0, 0);
+	daos_obj_generate_oid(coh, &oid, DAOS_OF_KV_FLAT, 0, 0, 0);
```
Comment (johannlombardi) [inline src/utils/daos_autotest.c:544]: ???

File: src/utils/daos_autotest.c [hunk 7]
```diff
@@ -578,7 +640,7 @@ kv_insert4k(void)
...
```
Comment (johannlombardi) [inline src/utils/daos_autotest.c:643]: ???

File: src/utils/daos_autotest.c [hunk 8]
```diff
@@ -641,7 +703,7 @@ kv_insert1m(void)
...
```
Comment (johannlombardi) [inline src/utils/daos_autotest.c:706]: ???

File: src/utils/daos_autotest.c [hunk 9]
```diff
@@ -760,11 +834,14 @@ struct step {
 };
 
 static struct step steps[] = {
+	/** Pre-test checks */
+	{ 0,	"Dumping AttachInfo",		attachinfo,	100 },
```
Comment (johannlombardi) [inline src/utils/daos_autotest.c:838]: This should verify that the agent is correctly running instead.

File: src/utils/daos_autotest.c [hunk 10]
```diff
@@ -714,12 +776,24 @@ static int
...
+	if (rc) {
+		step_fail(d_errdesc(rc));
+		return -1;
+	}
+
+	rc = daos_cont_destroy(poh, cuuid3, force, NULL);
+	if (rc) {
+		step_fail(d_errdesc(rc));
+		return -1;
+	}
```
Comment (johannlombardi) [inline src/utils/daos_autotest.c:795]: I am not sure what is the point of creating 2 extra containers with RF1 and 2 and not using them.

**Review by johannlombardi — CHANGES_REQUESTED** [review]
I really don't see the value of the patch in its current form. Sounds like we should review the original requirements.

**Comment by chrisphoffman** [comment]
> 
> 
> I really don't see the value of the patch in its current form. Sounds like we should review the original requirements.

To clarify is this comment meant for the test team to re-specify their requirements, directed at me or both?

**Comment by chrisphoffman** [comment]
> 
> 
> I think this needs rebased as there was a regression on master this morning.

This was rebased and it has passed the failing test. Thanks for the suggestion.

File: src/utils/daos_autotest.c (hunk 2 above)
Comment (chrisphoffman) [reply src/utils/daos_autotest.c:124]: That's because the command requires read permissions of agent.key file
-bash-4.2$ daos_agent dump-attachinfo
ERROR: daos_agent: Unable to load Certificate Data: could not load key: open /etc/daos/certs/agent.key: permission denied
logger Errorf() failed: write /tmp/daos_agent.log: file already closed

I also believe sudo is used to launch this same command in this other file:
src/tests/ftest/util/agent_utils.py
in this function:
    def dump_attachinfo(self):
        """Run dump-attachinfo on the daos_agent."""
        self.manager.job.set_sub_command("dump-attachinfo")
        self.manager.job.sudo = True

File: src/utils/daos_autotest.c (hunk 6 above)
Comment (chrisphoffman) [reply src/utils/daos_autotest.c:544]: That was my interpretation of how to use the API to auto-select the object class.
This file <> states:
" * \param[in]   cid     Class Identifier. This setting is for advanced users who
 *                      are knowledgeable on the specific oclass being set and
 *                      what that means for the object in the current system and
 *                      the container it's in.
 *                      Setting this to 0 (unknown) will check if there are any
 *                      hints specified and use an oclass accordingly. If there
 *                      are no hints specified we use the container properties
 *                      to select the object class."

I read that as setting cid and hints to 0 it will auto select object class from the container properties itself. This was attempted to be used to fulfill the requirement of "use the new API to auto-select the object class".

If my interpretation is incorrect can you let me know where to find this portion of the code?

**Comment by kprantis** [comment]
> > I really don't see the value of the patch in its current form. Sounds like we should review the original requirements.
> 
> To clarify is this comment meant for the test team to re-specify their requirements, directed at me or both?

I think it means that we need to have a conversation about what the requirements are, but not necessarily everyone will be watching the PR to see that, so I think the ask is probably for you to set something up to have a discussion about it.

File: src/utils/daos_autotest.c (hunk 6 above)
Comment (chrisphoffman) [reply src/utils/daos_autotest.c:544]: Discussed and this is fine.

File: src/utils/daos_autotest.c (hunk 1 above)
Comment (chrisphoffman) [reply src/utils/daos_autotest.c]: Labels are now utilized for containers.

File: src/utils/daos_autotest.c (hunk 3 above)
Comment (chrisphoffman) [reply src/utils/daos_autotest.c]: Labels are now utilized for containers.

File: src/utils/daos_autotest.c (hunk 5 above)
Comment (chrisphoffman) [reply src/utils/daos_autotest.c]: Labels are now utilized for containers.

File: src/utils/daos_autotest.c (hunk 10 above)
Comment (chrisphoffman) [reply src/utils/daos_autotest.c]: For RF1 and RF2 insert and read tests have been implemented.

File: src/utils/daos_autotest.c (hunk 4 above)
Comment (chrisphoffman) [reply src/utils/daos_autotest.c]: Labels are now utilized for containers.

File: src/utils/daos_autotest.c (hunk 2 above)
Comment (chrisphoffman) [reply src/utils/daos_autotest.c]: This is now removed. End users should not be running daos_agent in this fashion.

File: src/utils/daos_autotest.c (hunk 9 above)
Comment (chrisphoffman) [reply src/utils/daos_autotest.c]: This is addressed now in the daos command itself. 
When daos_agent isn't running it displays a meaningful message with a new errno:
```
daos ERR  src/common/drpc.c:217 unixcomm_connect() Failed to connect to /var/run/daos_agent/daos_agent.sock, errno=2(No such file or directory)
mgmt ERR  src/mgmt/cli_mgmt.c:234 get_attach_info() failed to connect to /var/run/daos_agent/daos_agent.sock DER_AGENT_COMM(-2034): 'Agent communication error'
ERROR: daos: daos_init() failed: DER_AGENT_COMM(-2034): Agent communication error
```

File: src/mgmt/cli_mgmt.c [hunk 11]
```diff
@@ -231,8 +231,8 @@ get_attach_info(const char *name, bool all_ranks, struct dc_mgmt_sys_info *info,
 	rc = drpc_connect(dc_agent_sockpath, &ctx);
 	if (rc != -DER_SUCCESS) {
 		D_ERROR("failed to connect to %s " DF_RC "\n",
-			dc_agent_sockpath, DP_RC(rc));
-		D_GOTO(out, 0);
+			dc_agent_sockpath, DP_RC(-DER_AGENT_COMM));
+		D_GOTO(out, rc = -DER_AGENT_COMM);
```
Comment (ashleypittman) [inline src/mgmt/cli_mgmt.c:235]: You need to only do this for the specific error that means the agent isn't running.
```suggestion
			dc_agent_sockpath, DP_RC(rc));
		if (rc == -DER_NOENT)
			rc = -DER_AGENT_COM;	
		D_GOTO(out, rc);
```

Comment (chrisphoffman) [reply src/mgmt/cli_mgmt.c:235]: This is how the output looks as stands in the PR:
`daos ERR  src/common/drpc.c:217 unixcomm_connect() Failed to connect to /var/run/daos_agent/daos_agent.sock, errno=2(No such file or directory)
mgmt ERR  src/mgmt/cli_mgmt.c:234 get_attach_info() failed to connect to /var/run/daos_agent/daos_agent.sock DER_AGENT_COMM(-2034): 'Agent communication error'
ERROR: daos: daos_init() failed: DER_AGENT_COMM(-2034): Agent communication error`

That suggestion yields:
`daos pool query --pool=8b15e064-c956-4bfa-ac46-64fb1f147f85
daos ERR  src/common/drpc.c:220 unixcomm_connect() Failed to connect to /var/run/daos_agent/daos_agent.sock, errno=2(No such file or directory)
mgmt ERR  src/mgmt/cli_mgmt.c:234 get_attach_info() failed to connect to /var/run/daos_agent/daos_agent.sock DER_NONEXIST(-1005): 'The specified entity does not exist'
ERROR: daos: daos_init() failed: DER_NONEXIST(-1005): The specified entity does not exist`

Is that the behavior we would like to see?

Comment (chrisphoffman) [reply src/mgmt/cli_mgmt.c]: This seems reasonable.

Comment (ashleypittman) [reply src/mgmt/cli_mgmt.c]: Having both -DER_NONEXIST and -DER_AGENT_COMM in the output actually makes the error clearer IMO, however I wasn't making my suggestion to improve the output, but saying that you shouldn't universally change the error code, just the  error codes that you expect for the failure you're trying to handle.  The fault injection tests (rightly) won't pass with the cod as it stands when I made my original comment.

Comment (ashleypittman) [reply src/mgmt/cli_mgmt.c]: Additionally, I don't have strong opinions on which should come first, the assignment or the D_ERROR() call, either would be fine depending on what you want the output to look like.

**Comment by kprantis** [comment]
I see all checks have passed - is this ready for people to take a look at again? Or still some prior feedback to be incorporated?

File: src/mgmt/cli_mgmt.c (hunk 11 above)
Comment (chrisphoffman) [reply src/mgmt/cli_mgmt.c]: I agreed with the comments a few days back. This is great feedback and I've incorporated these suggestions.

**Comment by chrisphoffman** [comment]
> 
> 
> I see all checks have passed - is this ready for people to take a look at again? Or still some prior feedback to be incorporated?

I'm looking for reviews from folks at this time.

File: src/include/daos_errno.h [hunk 12]
```diff
@@ -259,7 +259,10 @@ extern "C" {
 	       Fetch again)						\
 	/** Hit uncertain DTX, may need to try with other replica. */	\
 	ACTION(DER_TX_UNCERTAIN,	(DER_ERR_DAOS_BASE + 33),	\
-	       TX status is uncertaion)
+	       TX status is uncertaion)					\
+	/** Communicatin issue with agent. */				\
```
Comment (ashleypittman) [inline src/include/daos_errno.h:263]: Spelling error.

File: src/include/daos_errno.h [hunk 13]
```diff
@@ -259,7 +259,10 @@ extern "C" {
 	       Fetch again)						\
 	/** Hit uncertain DTX, may need to try with other replica. */	\
 	ACTION(DER_TX_UNCERTAIN,	(DER_ERR_DAOS_BASE + 33),	\
-	       TX status is uncertaion)
+	       TX status is uncertaion)					\
```
Comment (ashleypittman) [inline src/include/daos_errno.h:262]: uncertaion is also spelt wrong.

**Comment by kprantis** [comment]
Can we prioritize reviews on this PR please? It's one blocking feature freeze.

**Review by johannlombardi — APPROVED** [review]

**Review by ashleypittman — APPROVED** [review]
Other than the spelling mistakes in public header files I'm fine with this one.

It's a real shame that our CI turnaround takes so long so trivial issues like this cannot easily be fixed/resolved meaning we end up committing known-issues to the codebase.

**Review by bfaccini — APPROVED** [review]
